File: ios/adadapted_flutter_sdk/Sources/adadapted_flutter_sdk/AdadaptedFlutterSdkPlugin.swift
import AdSupport
import AppTrackingTransparency
import CoreTelephony
import Flutter
import UIKit

/// Gathers the device info the SDK reports with every event.
///
/// A port of the React Native SDK's `AdadaptedReactNativeSdk` module. The values here have
/// no Dart-only equivalent, which is why this package ships native code at all.
public class AdadaptedFlutterSdkPlugin: NSObject, FlutterPlugin {
    /// Must match the channel name the Dart side invokes.
    private static let channelName = "com.adadapted.flutter_sdk/device_info"

    /// The carrier reported when the device has none.
    private static let unknownCarrier = "n/a"

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: channelName,
            binaryMessenger: registrar.messenger()
        )

        registrar.addMethodCallDelegate(AdadaptedFlutterSdkPlugin(), channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard call.method == "getDeviceInfo" else {
            result(FlutterMethodNotImplemented)

            return
        }

        result(gatherDeviceInfo())
    }

    /// Collects everything the API wants to know about this device.
    private func gatherDeviceInfo() -> [String: Any] {
        let screenBounds = UIScreen.main.bounds
        let screenScale = UIScreen.main.scale
        let screenWidth = screenBounds.size.width * screenScale
        let screenHeight = screenBounds.size.height * screenScale

        let device = UIDevice.current
        let bundle = Bundle.main

        return [
            "udid": identifierForAdvertising(),
            "deviceName": device.model,
            // The platform suffix is how reporting tells a Flutter integration from a
            // React Native one. It must not be changed without the API being told.
            "systemName": "ios_flutter",
            "systemVersion": device.systemVersion,
            "deviceCarrier": carrierName(),
            "deviceModel": device.model,
            "deviceWidth": String(format: "%1.0f", screenWidth),
            "deviceHeight": String(format: "%1.0f", screenHeight),
            "deviceScreenDensity": String(format: "%0.0f", screenScale),
            "deviceLocale": Locale.preferredLanguages.first ?? "",
            "bundleId": bundle.bundleIdentifier ?? "",
            "bundleVersion": bundle.infoDictionary?["CFBundleShortVersionString"] as? String ?? "",
            "deviceTimezone": TimeZone.current.identifier,
            "isAdTrackingEnabled": isAdTrackingEnabled(),
        ]
    }

    /// The device's cellular carrier, or `n/a` when there is none.
    ///
    /// `serviceSubscriberCellularProviders` is deprecated from iOS 16 and Apple has said
    /// it will return `--` rather than a real name, so a missing or empty value is normal
    /// and is reported as `n/a`. Whatever it does return is forwarded as-is, placeholder
    /// included, because the native iOS and React Native SDKs do the same and reporting
    /// has to be able to compare iOS traffic across all three.
    private func carrierName() -> String {
        let networkInfo = CTTelephonyNetworkInfo()

        guard
            let carrier = networkInfo.serviceSubscriberCellularProviders?.values.first(where: {
                !($0.carrierName ?? "").isEmpty
            }),
            let name = carrier.carrierName,
            !name.isEmpty
        else {
            return Self.unknownCarrier
        }

        return name
    }

    /// Whether the user has permitted ad tracking.
    ///
    /// Via App Tracking Transparency, not `ASIdentifierManager` directly: Apple made
    /// `isAdvertisingTrackingEnabled` always return false from iOS 14, so reading it
    /// reports every user as having refused tracking, including those who granted it. This
    /// is the same signal that decides whether an advertising identifier is reported, so
    /// the two must agree.
    private func isAdTrackingEnabled() -> Bool {
        if #available(iOS 14, *) {
            return ATTrackingManager.trackingAuthorizationStatus == .authorized
        }

        return ASIdentifierManager.shared().isAdvertisingTrackingEnabled
    }

    /// The advertising identifier, or an empty string when tracking is not permitted.
    ///
    /// Nothing is substituted when the user has not permitted tracking. This matches the
    /// Android side, which leaves the identifier empty when the advertising ID is
    /// unavailable rather than reporting something else. `identifierForVendor` in
    /// particular is deliberately not used: it is stable and needs no prompt, but it is
    /// shared across this vendor's apps, and sending it to an ad service after tracking
    /// was denied is the sort of linkage App Tracking Transparency exists to prevent.
    private func identifierForAdvertising() -> String {
        guard isAdTrackingEnabled() else {
            return ""
        }

        return ASIdentifierManager.shared().advertisingIdentifier.uuidString
    }
}

File: android/src/main/kotlin/com/adadapted/flutter_sdk/AdadaptedFlutterSdkPlugin.kt
package com.adadapted.flutter_sdk

import android.content.Context
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.telephony.TelephonyManager
import android.util.DisplayMetrics
import android.util.Log
import com.google.android.gms.ads.identifier.AdvertisingIdClient
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.Locale
import java.util.TimeZone
import java.util.concurrent.Executors

/**
 * Gathers the device info the SDK reports with every event.
 *
 * A port of the React Native SDK's AdadaptedReactNativeSdkModule. The values here have no
 * Dart-only equivalent, which is why this package ships native code at all.
 */
class AdadaptedFlutterSdkPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var applicationContext: Context

    /**
     * Off the main thread, because AdvertisingIdClient.getAdvertisingIdInfo blocks on a
     * binder call to Play Services and throws IllegalStateException outright when called
     * from the main thread. The React Native module got away with running inline only
     * because the bridge already called it from a background thread; a Flutter method call
     * arrives on the platform thread, which is the main thread.
     */
    private val executor = Executors.newSingleThreadExecutor()

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        executor.shutdown()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (call.method != "getDeviceInfo") {
            result.notImplemented()

            return
        }

        executor.execute {
            val deviceInfo = gatherDeviceInfo()

            // Results must be delivered on the main thread, which is not the thread the
            // advertising ID was gathered on.
            android.os.Handler(android.os.Looper.getMainLooper()).post {
                result.success(deviceInfo)
            }
        }
    }

    /**
     * Collects everything the API wants to know about this device.
     */
    private fun gatherDeviceInfo(): Map<String, Any> {
        val displayMetrics: DisplayMetrics = applicationContext.resources.displayMetrics

        var gaid = ""
        var adTrackingEnabled = false
        var deviceCarrier = UNKNOWN_CARRIER
        var bundleVersion: String

        val telephonyManager =
            applicationContext.getSystemService(Context.TELEPHONY_SERVICE) as? TelephonyManager

        if (telephonyManager?.networkOperatorName != null) {
            deviceCarrier = telephonyManager.networkOperatorName
        }

        try {
            val gaidInfo = AdvertisingIdClient.getAdvertisingIdInfo(applicationContext)

            gaid = gaidInfo.id ?: ""
            adTrackingEnabled = !gaidInfo.isLimitAdTrackingEnabled
        } catch (ex: Exception) {
            // Play Services missing, out of date, or unavailable on this device. The
            // identifier is left empty rather than substituted, which is the same choice
            // the iOS side makes when tracking has not been permitted.
            Log.w(TAG, "Problem retrieving Google Play Advertiser Info", ex)
        }

        bundleVersion = try {
            val packageInfo: PackageInfo =
                applicationContext.packageManager.getPackageInfo(applicationContext.packageName, 0)

            // versionName is @Nullable on newer platform SDKs, so fall back to the same
            // value the NameNotFoundException branch reports rather than letting a null
            // reach the device payload.
            packageInfo.versionName ?: UNKNOWN_VALUE
        } catch (ex: PackageManager.NameNotFoundException) {
            UNKNOWN_VALUE
        }

        return mapOf(
            "udid" to gaid,
            "deviceName" to android.os.Build.DEVICE,
            // The platform suffix is how reporting tells a Flutter integration from a
            // React Native one. It must not be changed without the API being told.
            "systemName" to "android_flutter",
            "systemVersion" to android.os.Build.VERSION.RELEASE,
            "deviceCarrier" to deviceCarrier,
            "deviceModel" to android.os.Build.MODEL,
            "deviceWidth" to displayMetrics.widthPixels.toString(),
            "deviceHeight" to displayMetrics.heightPixels.toString(),
            "deviceScreenDensity" to displayMetrics.density.toString(),
            "deviceLocale" to Locale.getDefault().toString(),
            "bundleId" to applicationContext.packageName,
            "bundleVersion" to bundleVersion,
            "deviceTimezone" to TimeZone.getDefault().id,
            "isAdTrackingEnabled" to adTrackingEnabled,
        )
    }

    private companion object {
        /** Must match the channel name the Dart side invokes. */
        const val CHANNEL_NAME = "com.adadapted.flutter_sdk/device_info"

        const val TAG = "AdadaptedFlutterSdk"
        const val UNKNOWN_VALUE = "Unknown"
        const val UNKNOWN_CARRIER = "n/a"
    }
}
